### PR TITLE
Add skip and take

### PR DESCRIPTION
**Pull request by lhoestq**
As discussed in https://github.com/huggingface/datasets/pull/2375#discussion_r657084544 I added the `IterableDataset.skip` and `IterableDataset.take` methods that allows to do basic splitting of iterable datasets.

You can create new dataset with the first `n` examples using `IterableDataset.take()`, or you can get a dataset with the rest of the examples by skipping the first `n` examples with `IterableDataset.skip()`

One implementation detail:

Using `take` (or `skip`) prevents future dataset shuffling from shuffling the dataset shards, otherwise the taken examples could come from other shards. In this case it only uses the shuffle buffer.
I would have loved to allow the shards of the taken examples to be shuffled anyway, but since we don't know in advance the length of each shard we don't know what shards to take or skip.
I think this is ok though since users can shuffle before doing take or skip. I mentioned this in the documentation

cc @vblagoje @lewtun

### CODE REVIEW

File: src/datasets/iterable_dataset.py [hunk 1]
```diff
@@ -228,6 +241,43 @@ def n_shards(self) -> int:
...
+        self.n = n
+
+    def __iter__(self):
+        ex_iterator = iter(self.ex_iterable)
+        for _ in islice(ex_iterator, self.n):
+            pass
+        yield from ex_iterator
+
+    def shuffle_data_sources(self, seed: Optional[int]) -> "SkipExamplesIterable":
+        """Doesn't shuffle the wrapped examples iterable since it would skip exampels from other shards instead."""
```
Comment (vblagoje) [inline src/datasets/iterable_dataset.py:256]: typo: exampels

**Comment by vblagoje** [comment]
@lhoestq looks good. I tried with https://huggingface.co/datasets/vblagoje/wikipedia_snippets_streamed and it worked nicely. I would add more unit tests for edge cases. What happens if the n is larger than the total number of samples? Just to make sure these cases are handled properly.

**Comment by lhoestq** [comment]
Yup I'll add the tests thanks ;)

Moreover, I just noticed something in your wiki snippets code. FYI you're using `++ passage_counter ` at https://huggingface.co/datasets/vblagoje/wikipedia_snippets_streamed/blob/main/wikipedia_snippets_streamed.py#L102 but in python this doesn't increment the value @vblagoje

**Comment by vblagoje** [comment]
Thanks @lhoestq - not easy to convert after 10+ years of Java

File: docs/source/dataset_streaming.rst [hunk 2]
```diff
@@ -93,6 +96,29 @@ Tokenizers are written in Rust and use parallelism to speed up tokenization. To
     {'input_ids': [101, 11047, 10497, 7869, 2352...], 'token_type_ids': [0, 0, 0, 0, 0...], 'attention_mask': [1, 1, 1, 1, 1...]}
 
 
+Split your dataset with ``take`` and ``skip``
+--------------------------------------------------
+
+You can split your dataset by taking or skipping the first ``n`` examples.
+
+You can create new dataset with the first ``n`` examples using :func:`datasets.IterableDataset.take`, or you can get a dataset with the rest of the examples by skipping the first ``n`` examples with :func:`datasets.IterableDataset.skip`:
```
Comment (lewtun) [inline docs/source/dataset_streaming.rst:104]: nit: 

```suggestion
You can create a new dataset with the first ``n`` examples by using :func:`datasets.IterableDataset.take`, or you can create a dataset with the rest of the examples by skipping the first ``n`` examples with :func:`datasets.IterableDataset.skip`:
```

File: docs/source/dataset_streaming.rst [hunk 3]
```diff
@@ -93,6 +96,29 @@ Tokenizers are written in Rust and use parallelism to speed up tokenization. To
...
+
+
+.. code-block::
+
+    >>> dataset_head = dataset.take(2)
```
Comment (lewtun) [inline docs/source/dataset_streaming.rst:109]: would it make sense to include the `dataset = load_dataset(...)` line in the example?

File: docs/source/dataset_streaming.rst [hunk 4]
```diff
@@ -93,6 +96,29 @@ Tokenizers are written in Rust and use parallelism to speed up tokenization. To
...
+    >>> list(dataset_head)
+    [{'id': 0, 'text': 'Mtendere Village was...'}, '{id': 1, 'text': 'Lily James cannot fight the music...'}]
+    >>> # You can also create splits from a shuffled dataset
+    >>> train_dataset = shuffled_dataset.skip(1000)
+    >>> eval_dataset = shuffled_dataset.take(1000)
+
+Some things to keep in mind:
+
+- When you use ``skip``, on a dataset, then iterating on the new dataset will take some time to start. This is because under the hood it has to iterate over the skipped examples first.
```
Comment (lewtun) [inline docs/source/dataset_streaming.rst:118]: nit:

```suggestion
- When you apply ``skip`` to a dataset, iterating on the new dataset will take some time to start. This is because under the hood it has to iterate over the skipped examples first.
```

File: docs/source/dataset_streaming.rst [hunk 5]
```diff
@@ -93,6 +96,29 @@ Tokenizers are written in Rust and use parallelism to speed up tokenization. To
...
+- Using ``take`` (or ``skip``) prevents future calls to ``shuffle`` from shuffling the dataset shards order, otherwise the taken examples could come from other shards. In this case it only uses the shuffle buffer. Therefore it is advised to shuffle the dataset before splitting using ``take`` or ``skip``.
```
Comment (lewtun) [inline docs/source/dataset_streaming.rst:119]: is it possible to link to the location in the docs where "shuffle buffer" is defined? since it's a bit of a new concept (to me at least), we might want to help the user :)

File: src/datasets/iterable_dataset.py [hunk 6]
```diff
@@ -381,6 +437,40 @@ def shuffle(self, buffer_size, seed=None) -> "IterableDataset":
     def set_epoch(self, epoch: int):
         self._epoch = epoch
 
+    def skip(self, n) -> "IterableDataset":
+        """
+        Create a new IterableDataset that skips the first ``n`` elements.
+
+        Args:
+
```
Comment (lewtun) [inline src/datasets/iterable_dataset.py:445]: turbo nit: remove whitespace after `Args`

**Review by lewtun — COMMENTED** [review]
This feature looks really @lhoestq 🎉 

I left a few nits, but feel free to ignore them if you want
